### ISSUE TITLE
Lazy monkies arnt so lazy  

### ISSUE THREAD

**Issue by tonypee** [Yomguithereal/baobab#400]
It seems that the monkies are a little too fit. 
```
const tree2 = new Baobab({
  test: monkey(
      ['items'],
      function(items) {
        console.log('monkey eval');
      }
  ),
  items: []
});

var cursor = tree2.select('items');
cursor.push(1);
cursor.push(2);
cursor.push(3);

// output:  
// state.js:34 monkey eval 
// state.js:34 monkey eval
// state.js:34 monkey eval
```

So, the monkey is evaluated 3 times. It should only be on access no?

**Comment by SaphuA** [comment]
From the docs on the front page:

**Notes**

* The dynamic nodes will of course automatically update whenever at least one of the watched paths is updated.

So I'm pretty sure this is expected.

**Comment by markuplab** [comment]
"By access" is very bad way, because you can have many ui sources that use monkey fields.
But every mutation recalculation is also bad, i think it's can be recalculated on end of transactions (nextTick), or on sync read, if it's not cached.

**Comment by SaphuA** [comment]
@markuplab I'm pretty sure there is a system in place that prevents it from needlessly recalculating.

**Comment by Yomguithereal** [comment]
Okay, so let me explain what is the expected behavior of the monkeys:

* They should not compute before anything gets them.
* If they are gotten a second time, they won't recompute because they cached the computed value the first time.
* If a watched path is updated, they will be marked "dirty" and recompute next time they are gotten.

If this is not what you observe, and @tonypee's example seems to pinpoint this exactly, then there is a bug I must fix.

**Comment by SaphuA** [comment]
@Yomguithereal it seems that your last point...
* If a watched path is updated, they will be marked "dirty" and recompute next time they are gotten.

...is not the case in @tonypee's example. It seems to be recalculated every time a push to the watched path is done.

@tonypee can you provide a repro repo?

**Comment by Yomguithereal** [comment]
If the test case presented by @tonypee is full, no need for a repo. If this is all there is to such example then nothing should be logged to the console. So there indeed seems to be a bug.

**Comment by Yomguithereal** [comment]
Ok. I've pinpointed the source of the bug I think. Patch incoming.

**Comment by tonypee** [comment]
perfect - thanks!

**Comment by Yomguithereal** [comment]
Will need to fix some other things but a `2.2.0` is definitely coming soon.

**Comment by Yomguithereal** [comment]
Closing this then.
